### PR TITLE
[SPARK-37022][PYTHON] Use black as a formatter for  PySpark

### PR DESCRIPTION
**Pull request by zero323**
### What changes were proposed in this pull request?

This PR applies `black` (21.5b2) formatting to the whole `python/pyspark` source tree.

Additionally, the following changes were made:

- Disabled E501 (line too long) in pycodestyle config ‒ black allows line to exceed `line-length` in cases of inline comments.  There are 15 cases like this, all listed below
    ```
    pycodestyle checks failed:
    ./python/pyspark/sql/catalog.py:349:101: E501 line too long (103 > 100 characters)
    ./python/pyspark/sql/session.py:652:101: E501 line too long (106 > 100 characters)
    ./python/pyspark/sql/utils.py:50:101: E501 line too long (108 > 100 characters)
    ./python/pyspark/sql/streaming.py:1063:101: E501 line too long (128 > 100 characters)
    ./python/pyspark/sql/streaming.py:1071:101: E501 line too long (112 > 100 characters)
    ./python/pyspark/sql/streaming.py:1080:101: E501 line too long (124 > 100 characters)
    ./python/pyspark/sql/streaming.py:1259:101: E501 line too long (134 > 100 characters)
    ./python/pyspark/sql/pandas/conversion.py:136:101: E501 line too long (106 > 100 characters)
    ./python/pyspark/ml/param/_shared_params_code_gen.py:111:101: E501 line too long (103 > 100 characters)
    ./python/pyspark/ml/param/_shared_params_code_gen.py:136:101: E501 line too long (105 > 100 characters)
    ./python/pyspark/ml/param/_shared_params_code_gen.py:163:101: E501 line too long (101 > 100 characters)
    ./python/pyspark/ml/param/_shared_params_code_gen.py:233:101: E501 line too long (101 > 100 characters)
    ./python/pyspark/ml/param/_shared_params_code_gen.py:265:101: E501 line too long (101 > 100 characters)
    ./python/pyspark/tests/test_readwrite.py:235:101: E501 line too long (114 > 100 characters)
    ./python/pyspark/tests/test_readwrite.py:336:101: E501 line too long (114 > 100 characters)
    ```
- After reformatting, minor typing changes were made:
   - Realign certain `type: ignore` comments with ignored code.
   -  Apply explicit `casts` to  ` @unittest.skipIf` messages.  The following
      ```python
      @unittest.skipIf(
          not have_pandas or not have_pyarrow, pandas_requirement_message or pyarrow_requirement_message
      )  # type: ignore[arg-type]
      ```
      replaced with
              
      ```python
      @unittest.skipIf(
          not have_pandas or not have_pyarrow, 
          cast(str, pandas_requirement_message or pyarrow_requirement_message),
      )
      ```

### Why are the changes needed?

Consistency and reduced maintenance overhead.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Existing liners and tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
wow ..

**Comment by HyukjinKwon** [comment]
thanks for making a PR. cc @ueshin @xinrong-databricks @itholic FYI

**Review by srowen — COMMENTED** [review]
It's a good idea, but the amount of code churn is quite large (which causes back-port problems). Is there a subset of rules to enforce for now that requires much less change?

File: python/pyproject.toml [hunk 1]
```diff
@@ -23,3 +23,9 @@ testpaths = [
   "pyspark/sql/tests/typing",
   "pyspark/ml/typing",
 ]
+
+[tool.black]
+line-length = 100
+target-version = ['py37']
+include = '\.pyi?$'
+extend-exclude = 'cloudpickle'
```
[the remark this answers is not in the document]
Comment (zero323) [reply python/pyproject.toml:31]: To exclude `python/pyspark/python` that is already in `.gitignore`.

**Comment by HyukjinKwon** [comment]
I'm okay with this change. @zero323 would you mind resolving the conflicts?

**Comment by zero323** [comment]
> I'm okay with this change. @zero323 would you mind resolving the conflicts?

I'll take a look at this tomorrow. I assume we want to move `pyproject.toml` to `dev`, right?

**Comment by HyukjinKwon** [comment]
Yup, otherwise looks good.

File: python/pyspark/worker.py [hunk 2]
```diff
@@ -581,19 +627,18 @@ def main(infile, outfile):
             if bid >= 0:
                 if needs_broadcast_decryption_server:
                     read_bid = read_long(broadcast_sock_file)
-                    assert(read_bid == bid)
-                    _broadcastRegistry[bid] = \
-                        Broadcast(sock_file=broadcast_sock_file)
+                    assert read_bid == bid
+                    _broadcastRegistry[bid] = Broadcast(sock_file=broadcast_sock_file)
```
Comment (itholic) [inline python/pyspark/worker.py:631]: qq: Does black also fix this kind of changes automatically ??

**Comment by itholic** [comment]
Which version of `black` do you use for this changes ? We should use `21.5b2`.

Looks like it's correct version, but just for double checking just in case. :-)

**Review by itholic — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
PR description mentions `21.5b2` so we should be fine.

**Comment by HyukjinKwon** [comment]
I took a quick look, and I think this is fine to go ahead.

**Comment by HyukjinKwon** [comment]
im gonna merge this to avoid more conflicts. This is a big change but I believe this makes dev much easier ... let's see how it gose.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master!

**Comment by zero323** [comment]
Thanks @HyukjinKwon @itholic @srowen @xinrong-databricks!

**Comment by HyukjinKwon** [comment]
Forgot to cc other people like @BryanCutler @viirya who also work on PySpark side. Now we will apply auto-formatting via `dev/reformat-python`, and it will be verified via `dev/linter-python`.

For a bit of more background: this is a lot of changes but it boosts the development cycle given the experience in Databricks (e.g., Koalas) so we decided to get this in.

**Comment by BryanCutler** [comment]
SGTM +1

**Pull request closed without merging**